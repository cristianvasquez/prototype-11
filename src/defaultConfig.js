import { ns } from './namespaces.js'

// Things that can be configured, but only by a power user.

// returns: NamedNode
function pathToUri (path) {
  return ns.this[btoa(unescape(encodeURIComponent(path)))]
}

// Expects string of Named node
function uriToPath (uri) {
  const str = (typeof uri === 'string' || uri instanceof String) ? uri : uri.value
  const coding = str.replace(new RegExp(`^${ns.this()}`), '')
  return decodeURIComponent(escape(atob(coding)))
}

function selectToTable (sparqlSelectResult) {
  let header = null
  let rows = []
  for (const current of sparqlSelectResult) {
    if (!header) {
      header = Object.keys(current)
    }
    rows.push(Object.values(current).map((current) => current.value))
  }
  return {
    header: header,
    rows: rows
  }
}

function datasetToTable (dataset) {
  let header = ['subject', 'predicate', 'object']
  let rows = []
  // Map does not work?
  //   rows = dataset.toArray().map((quad)=>[quad.subject.value, quad.predicate.value, quad.object.value])
  for (const quad of dataset) {
    rows.push([quad.subject.value, quad.predicate.value, quad.object.value])
  }

  return {
    header: header,
    rows: rows
  }
}

const defaultConfig = {
  pathToUri,
  uriToPath,
  selectToTable,
  datasetToTable
}

export { defaultConfig }
