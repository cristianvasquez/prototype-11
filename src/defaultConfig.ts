import {ns} from './namespaces'
import {LINKS_REGEXP} from './lib/regexp'
import {TFile} from "obsidian";
import Dataset from "rdf-ext/lib/Dataset";
import {NamedNode, SparqlConfig} from "./types";

function replaceInternalLinks(query: string, getFirstLinkpathDest: any) {
    const convert = (str: any, p1: string, offset: any, s: any) => {
        const linkPath = getFirstLinkpathDest(p1)
        if (linkPath) {
            const uri = tFileToURI(linkPath).value
            return `<${uri}>`
        } else {
            throw new Error(`Cannot find:${p1}`)
        }
    }

    return query.replace(LINKS_REGEXP, convert)
}

function tFileToURI(tfile: TFile): NamedNode {
    // @ts-ignore
    return ns.this(btoa(unescape(encodeURIComponent(tfile.basename))))
}

function uriToNoteName(uri: NamedNode) {
    const coding = uri.value.replace(new RegExp(`^${ns.this()}`), '')
    return decodeURIComponent(escape(atob(coding)))
}

function isInternal(value: string) {
    //@ts-ignore
    return value.startsWith(ns.this())
}

function selectToTable(sparqlSelectResult: any) {
    let header: Array<string> = []
    let rows = []
    for (const current of sparqlSelectResult) {
        if (!header) {
            header = Object.keys(current)
        }
        rows.push(Object.values(current).map((current: NamedNode) => current.value))
    }
    return {
        header: header,
        rows: rows
    }
}

function datasetToTable(dataset: Dataset) {
    let header = ['subject', 'predicate', 'object']
    let rows = []
    // Map does not work?
    //   rows = dataset.toArray().map((quad)=>[quad.subject.value, quad.predicate.value, quad.object.value])
    for (const quad of dataset) {
        rows.push([quad.subject.value, quad.predicate.value, quad.object.value])
    }

    return {
        header: header,
        rows: rows
    }
}

const defaultConfig: SparqlConfig = {
    tFileToURI,
    uriToNoteName,
    replaceInternalLinks,
    isInternal,
    selectToTable,
    datasetToTable
}

export {defaultConfig}
