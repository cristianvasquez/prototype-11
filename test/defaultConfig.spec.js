import {defaultConfig} from '../src/defaultConfig.js'

import expect from 'expect'
import toMatchSnapshot from 'expect-mocha-snapshot'

expect.extend({ toMatchSnapshot })

describe('[uri]', function () {

  describe('[pathToUri]', function () {
    it(`"values"`, function () {
      const actual = defaultConfig.pathToUri('/hola/mundo')
      expect(actual).toMatchSnapshot(this)
    })
  })

  describe('[uriToPath]', function () {
    it(`"values"`, function () {
      const actual = defaultConfig.uriToPath('http://notes/L2hvbGEvbXVuZG8=')
      expect(actual).toMatchSnapshot(this)
    })
  })

  describe('[round]', function () {
    it(`"path-uri-path"`, function () {

      const uri = defaultConfig.pathToUri('/hola/mundo')
      const path = defaultConfig.uriToPath(uri)

      expect(path).toBe('/hola/mundo')
    })

    it(`"uri-path-uri"`, function () {

      const path = defaultConfig.uriToPath('http://notes/L2hvbGEvbXVuZG8=')
      const uri = defaultConfig.pathToUri(path)
      expect(uri.value).toBe('http://notes/L2hvbGEvbXVuZG8=')
    })

    it(`"emoji"`, function () {

      const uri = defaultConfig.pathToUri('/hola/üçé')
      const path = defaultConfig.uriToPath(uri)
      expect(path).toBe('/hola/üçé')
    })
  })

})