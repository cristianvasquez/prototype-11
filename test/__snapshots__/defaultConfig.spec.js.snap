// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[uri] [pathToUri] "values" 1`] = `
NamedNode {
  "value": "http://notes/L2hvbGEvbXVuZG8=",
}
`;

exports[`[uri] [uriToPath] "values" 1`] = `"/hola/mundo"`;
